### DISCUSSION TITLE
Possible memoryleak merging rows with clob values > 4k

### DISCUSSION THREAD

**Discussion by JesperSiteimprove** [godror/godror#133] [Q&A] [answered]
Hi

I am facing an issue that to me looks like a memoryleak somewhere behind the scenes of godror.
What my test program does, is to merge 200k rows in batches of 10k rows, into a table with just a number and a clob column.
Each inserted string is 75 bytes of the character "a". Id is just from 0-200k.
Running the program in this "version", I do not see any sign of increasing memory usage.

If I then change to program to instead insert 20k bytes of the character "a" but only for the first row, and still inserting 75 bytes of the character "a" in the remaining 199.999 rows, then the memory usage keeps growing.

For each iteration of merging the 200k rows the program writes the memstats like this:
Alloc = 28 MiB	TotalAlloc = 3388 MiBHeapInUse = 29 MiB	Sys = 71 MiB	NumGC = 421

But I dont see any increases in those memstats, besides the TotalAlloc, which is expected.

Monitoring the running program using htop, I see the RES growing.

Hope you can help

Best Regards
Jesper

**Comment by tgulacsi** [comment]
Maybe the memory increase is in the ODPI/OCI/C library. Can you give the two code samples as test functions?

**Comment by JesperSiteimprove** [comment]
Run the 2 tests like this:

-- first test is stable in memory
FIRST_ROW_BYTES=75 ORA_CONN_STRING=USER/PW@DBSID go test -v -run TestMergeMemory

-- second test has increasing memory
FIRST_ROW_BYTES=20000 ORA_CONN_STRING=USER/PW@DBSID go test -v -run TestMergeMemory

```
package main

import (
	"bufio"
	"bytes"
	"context"
	"database/sql"
	"fmt"
	"math"
	"os"
	"runtime"
	"strconv"
	"strings"
	"testing"
	"time"

	"github.com/godror/godror"
)

func TestMergeMemory(t *testing.T) {
	dbConnectString := os.Getenv("ORA_CONN_STRING")
	firstRowBytes, err := strconv.Atoi(os.Getenv("FIRST_ROW_BYTES"))
	if err != nil {
		t.Fatalf("Must set env variable FIRST_ROW_BYTES, %s", err)
	}

	params, err := godror.ParseConnString(dbConnectString)
	if err != nil {
		t.Fatal(err)
	}
	params.StandaloneConnection = true
	db, err := sql.Open("godror", params.StringWithPassword())
	if err != nil {
		t.Fatal(err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()

	conn, err := db.Conn(ctx)
	if err != nil {
		t.Fatal(err)
	}

	if _, err := conn.ExecContext(ctx, "DROP TABLE test_many_rows"); err != nil {
		t.Fatal(err)
	}
	if _, err := conn.ExecContext(ctx, "CREATE TABLE test_many_rows (id number(10), text clob)"); err != nil {
		t.Fatal(err)
	}
	if _, err := conn.ExecContext(ctx, "CREATE index test_many_rows_id_idx on test_many_rows(id)"); err != nil {
		t.Fatal(err)
	}

	loopCnt := 0
	for {
		loopCnt++
		ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
		defer cancel()

		tx, err := conn.BeginTx(ctx, nil)
		if err != nil {
			t.Fatalf("Failed to beginTx: %s\n", err)
		}
		defer tx.Rollback()

		stmt, err := tx.PrepareContext(ctx, `MERGE /*+index(x)*/ INTO test_many_rows x 
			USING (select  :ID ID, :text text from dual) y 
			ON (x.id = y.id ) 
			WHEN MATCHED 
				THEN UPDATE SET  x.text = y.text 
			WHEN NOT MATCHED 
				THEN INSERT ( x.ID, x.TEXT) VALUES ( y.ID, y.TEXT)`)

		if err != nil {
			t.Errorf("Failed to prepare insSQL %s", err)
			_ = stmt.Close()

		}
		defer stmt.Close()

		rowsToInsert := 200000
		mergeIds := make([]sql.NullInt64, 0)
		mergeStrings := make([]string, 0)
		totalBytes := 0
		for i := 0; i != rowsToInsert; i++ {
			byteLen := 75
			if i < 1 {
				byteLen = firstRowBytes
				fmt.Printf("first row bytes len: %d\n", byteLen)
			}
			str := strings.Repeat("a", byteLen)
			mergeStrings = append(mergeStrings, str)
			totalBytes += len(str)
			mergeIds = append(mergeIds, sql.NullInt64{Int64: int64(i), Valid: true})
		}
		fmt.Printf("leng(mergeIds): %d, len(mergeSTrings): %d, total size of clob strings: %d MiB\n", len(mergeIds), len(mergeStrings), totalBytes/1024/1024)

		var maxValuesPerOp int = 10000
		maxLoopCount := int(math.Ceil(float64(rowsToInsert) / float64(maxValuesPerOp)))
		totalMilliseconds := 0
		totalRowsMerged := 0
		fmt.Printf("Start merge loopCnt: %d\n", loopCnt)
		for loopIdx := 0; loopIdx < maxLoopCount; loopIdx++ {
			sliceStartIdx := loopIdx * maxValuesPerOp
			sliceEndIdx := int(rowsToInsert)
			if sliceStartIdx+maxValuesPerOp < int(rowsToInsert) {
				sliceEndIdx = sliceStartIdx + maxValuesPerOp
			}

			beforeExecTs := time.Now()
			var res sql.Result
			binds := make([]interface{}, 0)

			int64Elements := make([]sql.NullInt64, sliceEndIdx-sliceStartIdx)
			copy(int64Elements, mergeIds[sliceStartIdx:sliceEndIdx])
			binds = append(binds, sql.Named("id", int64Elements))
			stringElements := make([]string, sliceEndIdx-sliceStartIdx)
			copy(stringElements, mergeStrings[sliceStartIdx:sliceEndIdx])
			binds = append(binds, sql.Named("text", stringElements))

			res, err = stmt.ExecContext(ctx, binds...)
			if err != nil {
				tx.Rollback()
				stmt.Close()
				binds = nil
				t.Errorf("Failed to execute: %s\n", err)
			}

			totalMilliseconds += int(time.Now().Sub(beforeExecTs).Milliseconds())

			binds = nil

			rowsAffected, err := res.RowsAffected()
			if err != nil {
				t.Errorf("problem getting res.RowsAffected(): %s\n", err)
			}
			totalRowsMerged += int(rowsAffected)
		}
		tx.Commit()
		mergeIds = nil
		mergeStrings = nil
		fmt.Printf("Merge done, number of rows merged: %d, merge timing msec: %d\n", totalRowsMerged, totalMilliseconds)

		stmt.Close()
		if err != nil {
			t.Errorf("problem with stmt.Close: %s\n", err)
		}

		var m runtime.MemStats
		runtime.ReadMemStats(&m)
		fmt.Printf("Alloc = %v MiB", m.Alloc/1024/1024)
		fmt.Printf("\tTotalAlloc = %v MiB", m.TotalAlloc/1024/1024)
		fmt.Printf("HeapInUse = %v MiB", m.HeapInuse/1024/1024)
		fmt.Printf("\tSys = %v MiB", m.Sys/1024/1024)
		fmt.Printf("\tNumGC = %v\n", m.NumGC)

		f, err := os.Open(fmt.Sprintf("/proc/%d/smaps", os.Getpid()))
		if err != nil {
			t.Fatal(err)
		}
		defer f.Close()

		pss := uint64(0)
		pfx := []byte("Pss:")
		r := bufio.NewScanner(f)
		for r.Scan() {
			line := r.Bytes()
			if bytes.HasPrefix(line, pfx) {
				var size uint64
				_, err := fmt.Sscanf(string(line[4:]), "%d", &size)
				if err != nil {
					t.Fatal(err)
				}
				pss += size
			}
		}
		if err := r.Err(); err != nil {
			t.Fatal(err)
		}

		fmt.Printf("Current process memory (Pss): %d MiB\n", pss/1024)

	}
}

```

**Comment by sudarshan12s** [comment]
few things I noticed.

- defer statements are used inside a loop which will never get executed as
they are run only on function exit
-  conn.Close needs to be added
- binds = nil // it would cause GC not to cleanup this variable binds


checking more on reproducing the issue.

Thanks

**Reply by JesperSiteimprove** [reply]
The original test program had multiple functions, but when I made the test function yesterday, I just put all the code into one function, so that explains the defer statements.
The test function never stops so I guess it doesn't matter that conn.Close is not called.
Setting variables to nil, I actually thought that you helped the GC to clean up earlier? But again, maybe it doesn't matter when there is only one function that never exits.
Finally the memstats does not increase, so I dont think it is any of variables in the test function that builds up the memory.

Thanks for looking at the issue

**Comment by tgulacsi** [comment]
With 10000 rows inserted, with `FIRST_ROW_BYTES=8191`, the Pss stays around 50MiB.
With `FIRST_ROW_BYTES=8192`, the Pss increments about 32MiB per round.

TestMergeMemory133 at 970f5cd.

**Reply by JesperSiteimprove** [reply]
I am so glad that you were able to reproduce the issue in your end. I have been going nuts trying to find this memory leak in my production program, until one day where I tore the program to pieces and found out that it was something with clob columns and finally that the size of the clob somehow was the culprit.

Thanks again for spending time on this

**Comment by tgulacsi** [comment]
Even stranger: with `ROWS_TO_INSERT=1000`, the memory is around 100MiB; with `ROWS_TO_INSERT=10000` (10x), the memory is around 40MiB.

**Comment by tgulacsi** [comment, marked as answer]
You use a slice to insert (batch insert).
Such arrays requires to have all elements be the same size (a continuous memory chunk is allocated, to which all elements fit into).
I think that's why the first element's size has such a substantial effect.

The next question is that why aren't those bytes are freed.

With using `godror.Lob{Reader:strings.NewReader(...)}` I don't see such memory usage.

**Reply by JesperSiteimprove** [reply]
Ufortunately using `godror.Lob` inserting is very much slower.

I have now tried to use `godror.Lob` and for each loop I still see an increase in memory.

First Pss is 27.640 MiB and the last, before the test run stops after 10 minutes and 38 loops, is 230.546 MiB.
Definitely a much slower memory increase, but not stable it seems.

**Reply by tgulacsi** [reply]
PTAL bb9c38321f358af8286e70b52bff2259981ae22b - I've simplified your test, as I couldn't make my head wrap around it.
But this shrunk the memory inflation.

So, can you use TestMergeMemory in `z_issue133_test.go`, or how to modify it to resemble more your use case?
We want a minimal reproducer, if possible.

**Reply by JesperSiteimprove** [reply]
I think your rewrite of the test still looks correct, but the thing that now causes it not to leak, is the fact that you set the batchsize to 100 instead of 10.000.
If I set the constant to 10k (const batchSize = 10000) and call the test with these parameters:
`ROWS_TO_INSERT=200000 USE_LOBS=false FIRST_ROW_BYTES=20000`
then the memory is at 2,4 GB after 30 loops.

Setting the batchsize that low equals to a lot of roundtrips to the database which is not optimal, when you need to insert/update a high number of rows.

**Reply by tgulacsi** [reply]
I've written a test-runner test that uses `DPI_DEBUG_LEVEL=32` to check whether this is an ODPI error.

    go test -run=Alloc133 -v 2>&1 | tee /tmp/alloc133.txt

But the unfreed area is only half a megabyte:
```
echo '
Remaining allocations: map[2442117888:OCI allocated 8 bytes at 0x19f83c0 2701237760:OCI allocated 264 bytes at 0x1a10e90 2701803008:OCI allocated 4016 bytes at 0x1aa0e30 2702847488:OCI allocated 4160 bytes at 0x1aa1e20 2703924736:OCI allocated 12352 bytes at 0x1aa2e90 2703996160:OCI allocated 4160 bytes at 0x1ab25b0 2704379136:OCI allocated 32 bytes at 0x1a13d80 2704999936:OCI allocated 4160 bytes at 0x1ab3600 2706068992:OCI allocated 4160 bytes at 0x1ab4650 2706196736:OCI allocated 29056 bytes at 0x1ad4940 2707090944:OCI allocated 12352 bytes at 0x1aa5ee0 2707138048:OCI allocated 4160 bytes at 0x1ab56a0 2708207104:OCI allocated 4160 bytes at 0x1ab66f0 2708836096:allocated 64 bytes at 0x1a57f80 (allocate handle list) 2708856576:OCI allocated 4016 bytes at 0x1a57fd0 2709210880:OCI allocated 4160 bytes at 0x1ab7740 2709888768:allocated 56 bytes at 0x1a58f90 (allocate handle) 2710191872:OCI allocated 4672 bytes at 0x1aa8f30 2710279936:OCI allocated 7296 bytes at 0x1ab8790 2710900224:allocated 168 bytes at 0x1a59e00 (allocate handle) 2710913024:allocated 64 bytes at 0x1a59040 (allocate handle list) 2710924544:allocated 6 bytes at 0x1a59d60 (allocate default encoding) 2710932736:allocated 17 bytes at 0x1a59d80 (allocate default driver name) 2710946048:OCI allocated 168 bytes at 0x1a591c0 2711382016:OCI allocated 4208 bytes at 0x1ac9860 2712193024:OCI allocated 4208 bytes at 0x1a8a8c0 2712305920:OCI allocated 4744 bytes at 0x1aaa180 2712347648:OCI allocated 4208 bytes at 0x1aba420 2712463360:OCI allocated 8344 bytes at 0x1aca8e0 2713209088:OCI allocated 4208 bytes at 0x1a8b940 2713326592:OCI allocated 4160 bytes at 0x1aab410 2713428992:OCI allocated 4208 bytes at 0x1abb4a0 2714054656:OCI allocated 59368 bytes at 0x1a5c030 2714290432:allocated 3216 bytes at 0x1a8c9c0 (allocate error buffer) 2714355968:OCI allocated 4016 bytes at 0x1a9c9c0 2714395648:OCI allocated 4160 bytes at 0x1aac460 2714445056:OCI allocated 4208 bytes at 0x1abc520 2715289600:allocated 64 bytes at 0x1a8d800 (allocate handle list slots) 2715298048:OCI allocated 59368 bytes at 0x1a8d920 2715310080:allocated 64 bytes at 0x1a8d850 (allocate handle list) 2715313664:allocated 328 bytes at 0x1a8d660 (allocate env) 2715330560:allocated 64 bytes at 0x1a8d8a0 (allocate handle list slots) 2715334400:allocated 64 bytes at 0x1a8d7b0 (allocate handle list slots) 2715351040:OCI allocated 32 bytes at 0x1a8d8f0 2715396352:OCI allocated 13232 bytes at 0x1a9d9a0 2715472896:OCI allocated 4160 bytes at 0x1aad4d0 2715514368:OCI allocated 4160 bytes at 0x1abd670 2716476672:OCI allocated 4160 bytes at 0x1aae520 2716538624:allocated 64 bytes at 0x1abe710 (allocate handle pool slots) 2716559104:OCI allocated 4208 bytes at 0x1abe760 2716583424:allocated 64 bytes at 0x1abe6c0 (allocate handle pool) 2717545728:OCI allocated 12336 bytes at 0x1aaf570 2717600000:OCI allocated 40280 bytes at 0x1abf940 2969948928:OCI allocated 4328 bytes at 0x1b503d0 2970993664:OCI allocated 4208 bytes at 0x1b514c0 2974110720:OCI allocated 4256 bytes at 0x1b54450 2975156224:OCI allocated 4256 bytes at 0x1b55840 2979155456:allocated 89 bytes at 0x1b29e40 (allocate release string) 271584992473094:allocated 3216 bytes at 0x7f10f4000b60 (allocate error buffer) 271587123179526:allocated 3216 bytes at 0x7f10ec000b60 (allocate error buffer) 271587139960844:allocated 3216 bytes at 0x7f10fc000cc0 (allocate error buffer) 271652386549766:allocated 3216 bytes at 0x7f1100000b60 (allocate error buffer) 271652420108300:allocated 3216 bytes at 0x7f1120000cc0 (allocate error buffer) 271653460295692:allocated 3216 bytes at 0x7f1104000cc0 (allocate error buffer) 271653493862412:allocated 3216 bytes at 0x7f1124000fc0 (allocate error buffer) 271653512808961:OCI allocated 178360 bytes at 0x7f113422a010 271654567591945:allocated 3216 bytes at 0x7f1128000c90 (allocate error buffer) 271655624552454:allocated 3216 bytes at 0x7f111c000b60 (allocate error buffer)]
' | tr ':' '\n' | grep -o ' [0-9]* bytes' | awk '{s+=$1} END{print s}
```
584072

[alloc133.txt.gz](https://github.com/godror/godror/files/6021056/alloc133.txt.gz)

**Reply by tgulacsi** [reply]
```
$ go test -run=Alloc133 -v

=== RUN   TestMemoryAlloc133
    z_issue133_test.go:41: Start separate test process of env DPI_DEBUG_LEVEL=32 RUNS=2 [./godror.test -test.run=^TestMergeMemory133$ -test.count=1 -test.v]
    z_issue133_test.go:61:     z_issue133_test.go:178: first row bytes len: 20000
    z_issue133_test.go:61:     z_issue133_test.go:183: Start merge loopCnt: 0
    z_issue133_test.go:61:     z_issue133_test.go:227: Total size of clob strings: 15019925
    z_issue133_test.go:61:     z_issue133_test.go:273: Merge done, number of rows merged: 200000, merge timing: 24.476682193s
    z_issue133_test.go:61:     z_issue133_test.go:188: Alloc: 0.857 MiB, Heap: 1.508 MiB, Sys: 71.079 MiB, NumGC: 124
    z_issue133_test.go:61:     z_issue133_test.go:195: Current process memory (Pss): 530.8975 MiB
    z_issue133_test.go:61:     z_issue133_test.go:183: Start merge loopCnt: 1
    z_issue133_test.go:61:     z_issue133_test.go:227: Total size of clob strings: 15019925
    z_issue133_test.go:61:     z_issue133_test.go:273: Merge done, number of rows merged: 200000, merge timing: 17.900174284s
    z_issue133_test.go:61:     z_issue133_test.go:188: Alloc: 2.531 MiB, Heap: 3.164 MiB, Sys: 71.079 MiB, NumGC: 242
    z_issue133_test.go:61:     z_issue133_test.go:195: Current process memory (Pss): 1048.2842 MiB
    z_issue133_test.go:101: Remained 584072 bytes in 68 allocations.
--- PASS: TestMemoryAlloc133 (44.89s)
```

The strange thing here is the difference between Pss (1GiB) and Sys (Total allocated memory by Go runtime: 71MiB).

**Reply by JesperSiteimprove** [reply]
It seems very strange indeed.

This guy talks about some different tools that he used to find what was leaking in cgo:
https://kirshatrov.com/2019/11/04/finding-memory-leak-in-cgo/
Dont know if that could somehow help?

Best Regards
Jesper

**Reply by tgulacsi** [reply]
Heaptrack:
```
200118 calls to allocation functions with 1.31GB peak consumption from
dpiUtils__allocateMemory
  at /home/gthomas/src/github.com/godror/godror/odpi/embed/../src/dpiUtils.c:29
  in /home/gthomas/src/github.com/godror/godror/godror.test
200000 calls with 1.31GB peak consumption from:
    dpiVar__allocateDynamicBytes
      at /home/gthomas/src/github.com/godror/godror/odpi/embed/../src/dpiVar.c:179
      in /home/gthomas/src/github.com/godror/godror/godror.test
    dpiVar__setFromBytes
      at /home/gthomas/src/github.com/godror/godror/odpi/embed/../src/dpiVar.c:1245
      in /home/gthomas/src/github.com/godror/godror/godror.test
    dpiVar_setFromBytes
      at /home/gthomas/src/github.com/godror/godror/odpi/embed/../src/dpiVar.c:1774
      in /home/gthomas/src/github.com/godror/godror/godror.test
    runtime.asmcgocall
      at /usr/local/go/src/runtime/asm_amd64.s:667
      in /home/gthomas/src/github.com/godror/godror/godror.test
40 calls with 80.00KB peak consumption from:
    dpiVar__initBuffer
      at /home/gthomas/src/github.com/godror/godror/odpi/embed/../src/dpiVar.c:883
      in /home/gthomas/src/github.com/godror/godror/godror.test
    dpiVar__allocate
      at /home/gthomas/src/github.com/godror/godror/odpi/embed/../src/dpiVar.c:118
      in /home/gthomas/src/github.com/godror/godror/godror.test
    dpiConn_newVar
      at /home/gthomas/src/github.com/godror/godror/odpi/embed/../src/dpiConn.c:2023
      in /home/gthomas/src/github.com/godror/godror/godror.test
    _cgo_1c0b3c4c822f_Cfunc_dpiConn_newVar
      at /tmp/go-build/cgo-gcc-prolog:122
      in /home/gthomas/src/github.com/godror/godror/godror.test
    runtime.asmcgocall
      at /usr/local/go/src/runtime/asm_amd64.s:667
      in /home/gthomas/src/github.com/godror/godror/godror.test
```

When I count the ODPI-reported allocations and frees, it's even stranger, as that says 6.5GiB allocated and *freed*. But pmap says RSS raises about 200MiB per run.
```
=== RUN   TestMemoryAlloc133
    z_issue133_test.go:46: Start separate test process of env DPI_DEBUG_LEVEL=32 RUNS=3 [./godror.test -test.run=^TestMergeMemory133$ -test.count=1 -test.v]
    z_issue133_test.go:96:     z_issue133_test.go:215: ---BEGIN merge loopCnt: 0
    z_issue133_test.go:96:     z_issue133_test.go:269: Total size of clob strings: 7519925
    z_issue133_test.go:96:     z_issue133_test.go:315: Merge done, number of rows merged: 100000, merge timing: 16.273333305s
    z_issue133_test.go:96:     z_issue133_test.go:226: Alloc: 0.811 MiB, Heap: 1.562 MiB, Sys: 70.579 MiB, NumGC: 68
    z_issue133_test.go:88:     z_issue133_test.go:233: ---END merge loopCnt: 0; process memory (Pss): 368.145 MiB
    z_issue133_test.go:65: Remained 12864 bytes from 6581096470 in 4 allocations.
    z_issue133_test.go:96:     z_issue133_test.go:215: ---BEGIN merge loopCnt: 1
    z_issue133_test.go:96:     z_issue133_test.go:269: Total size of clob strings: 7519925
    z_issue133_test.go:96:     z_issue133_test.go:315: Merge done, number of rows merged: 100000, merge timing: 9.835074413s
    z_issue133_test.go:96:     z_issue133_test.go:226: Alloc: 0.812 MiB, Heap: 1.555 MiB, Sys: 70.579 MiB, NumGC: 125
    z_issue133_test.go:88:     z_issue133_test.go:233: ---END merge loopCnt: 1; process memory (Pss): 572.895 MiB
    z_issue133_test.go:65: Remained 12864 bytes from 6581096470 in 4 allocations.
    z_issue133_test.go:96:     z_issue133_test.go:215: ---BEGIN merge loopCnt: 2
    z_issue133_test.go:96:     z_issue133_test.go:269: Total size of clob strings: 7519925
    z_issue133_test.go:96:     z_issue133_test.go:315: Merge done, number of rows merged: 100000, merge timing: 10.592547284s
    z_issue133_test.go:96:     z_issue133_test.go:226: Alloc: 0.814 MiB, Heap: 1.578 MiB, Sys: 70.579 MiB, NumGC: 185
    z_issue133_test.go:88:     z_issue133_test.go:233: ---END merge loopCnt: 2; process memory (Pss): 785.754 MiB
    z_issue133_test.go:65: Remained 12864 bytes from 6581096470 in 4 allocations.
```

**Reply by tgulacsi** [reply]
As I don't see any memory leak, I've tried to eliminate fragmentation:

with out-of-the-box glibc, memory increases by 200MiB per iteration

    go test -run=Memory133$ -v

with tcmalloc, memory increases in each iteration by 4-8MiB

    LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4.5.6 go test -run=Memory133$ -v

with jemalloc, memory stays around 43MiB +-0.5MiB.

    LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.1 go test -run=Memory133$ -v

So I see no evicence of memory leak, but fragmentation.

**Reply by JesperSiteimprove** [reply]
ok - thanks for the update.
What does that exactly mean? Is that a solution, something that godror can change internally?

Best Regards
Jesper

**Reply by JesperSiteimprove** [reply]
I can confirm that setting `LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.1` before starting the test, memory usage is very low and very stable.
Is that just something I should use when I run my programs or do you see it as something godror should always use as default?

**Reply by tgulacsi** [reply]
This means that this allocation pattern (allocate batchSize * 65536 bytes, release them, and repeat) is pathologic for glibc.
It is handled well by jemalloc.

But I've found a half-solution: reuse.
If we don't release the dpiVar structure, then it can keep its allocated *dpiData structures, so less new allocations is needed.

Usually we do keep dpiVar, but not every time.
I've implemented a more fine grained logic for this. PTAL 093fc2e44f09b05beeb5b0bbf766be623f3e0cfb !

**Reply by JesperSiteimprove** [reply]
Hi Tamás

Thanks for all the time you spend on this issue. I really appreciate it.

I have tried to put you modified stmt.go in my mod files (not sure how I use a branch version in an elegant way) and run the previous version of `issue133Inner`, and after 30 loops the Pss is at 782 MiB, compared to my previous post with 2,4 GB . So a lot better, but not nearly as good as with your `libjemalloc` suggestion, where the Pss is around 29 MiB after 30 loops.

Do you know of any downsides to just always use the jemalloc?

**Reply by tgulacsi** [reply]
I've released v0.24.0.

Regarding jemalloc: on an 8 processor machine I've had to lower GOMAXPROCS to 4 otherwise it failed malloc-ing.
So your miles may wary - test it.

**Comment by sudarshan12s** [comment]
Jesper, does workarounds like separating merges for strings of size >32k or
20k in your case  using LOB reader work for you?

GOGC=1, debug.FreeOSMemory seems to do some cleanup of the Go runtime
memory and returns it to OS, thus reducing the RES footprint.
I see Go runtime starts releasing memory to OS  even if another program (an
intentional test program) starts consuming high memory.

there was still some RES memory which is  not clear.

Thanks

**Reply by JesperSiteimprove** [reply]
Hi Sudarshan
I have a considered a workaround like you propose, but haven't gotten around to trying to implement that. It seems like kind of a minefield and would rather have a general solution that others could also benefit from.
Best Regards
Jesper

**Reply by tgulacsi** [reply]
For big BLOBs use of godror.Lob would be better.
But nowadays a 1MiB LOB is not big.
We should handle such use case (10_000 rows batch insert).

**Reply by tgulacsi** [reply]
Go 1.16 has returned to use of MADV_DONTNEED (https://tip.golang.org/doc/go1.16#runtime) so such FreeOSMemory should be uneffective. Also, here the Go memory usage stays around 70MiB, all other os on C (ODPI) side.

**Comment by sudarshan12s** [comment]
Agreed. Yes as you mentioned it is  related to fragmentation due to C
malloc/frees and not with Go memory by default allocator.

I ran it through valgrind without your latest patch on reducing
allocations. process RSS memory stay around 500MiB  and valgrind.out has no
major leaks in summary. without valgrind , it cross 1.5 G as you saw it
with default glibc allocator.
go test -c -o testissue133
valgrind --log-file=valgrind.out  --leak-check=full  --show-leak-kinds=all
  ./testissue133 -test.v -test.run=Memory133$
  z_issue133_test.go:235: ---END merge loopCnt: 0; test process memory
(rss): 496.578 MiB
z_issue133_test.go:235: ---END merge loopCnt: 1; test process memory (rss):
502.227 MiB
...

  z_issue133_test.go:235: ---END merge loopCnt: 63; test process memory
(rss): 548.086 MiB

mostly valgrind override of malloc might have reduced the fragmentation.
